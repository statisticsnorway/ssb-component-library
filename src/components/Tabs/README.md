Tabs
========

> A tab selector component

### Usage

#### HTML

```html
<div class="ssb-tabs">
    <button class="navigation-item ">
        <span>Item 1</span>
    </button>
    <button class="navigation-item ">
        <span>Item 2</span>
    </button>
    <button class="navigation-item ">
        <span>Item 3</span>
    </button>
</div>
```

#### React

```jsx harmony
const items = [
  {
    title: 'Item 1',
    path: 'item1/',
  },
  {
    title: 'Item 2',
    path: 'item2/',
  },
  {
    title: 'Item 3',
    path: 'item3/',
  },
];

<Tabs items={items} />
```

Available props:

| Name       | Type          | Description  |
| ---------- | ------------- | ----- |
| activeOnInit | string | Will set an item to be active on init |
| items | arrayOf({title, path}) | Sets label and path of buttons |
| onClick | func | Callback function |
