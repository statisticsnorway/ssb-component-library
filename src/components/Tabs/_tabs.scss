.ssb-tabs {
  @include roboto;
  display: flex;
  flex-direction: row;

  .navigation-item {
    @include reset-button;
    align-items: center;
    cursor: pointer;
    display: flex;
    justify-content: center;
    padding: 10px;
    position: relative;
    text-align: center;
    text-decoration: none;
    transition: background .2s;

    &:hover {
      background: $ssb-green-2;
    }

    &:before {
      background: $ssb-green-3;
      bottom: 0;
      content: '';
      display: block;
      height: 0;
      left: 0;
      position: absolute;
      transition: height .2s;
      width: 100%;
    }

    &.active {

      &:before {
        height: 5px;
      }
    }
  }
}
