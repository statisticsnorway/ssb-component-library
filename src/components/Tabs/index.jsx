import React, { useState } from 'react';
import PropTypes from 'prop-types';

const Tabs = ({ activeOnInit, className, items, onClick }) => {
	const [activeItem, changeActive] = useState(activeOnInit);

	const handleClick = e => {
		onClick(e);
		changeActive(e);
	};

	return (
		<div className={`ssb-tabs${className ? ` ${className}` : ''}`}>
			{items.map(item => (
				<button
					className={`navigation-item ${activeItem === item.path ? 'active' : ''}`}
					onClick={() => handleClick(item.path)}
					key={item.path}
				><span>{item.title}</span>
				</button>
			))}
		</div>
	);
};

Tabs.defaultProps = {
	activeOnInit: '',
	onClick: () => {},
};

Tabs.propTypes = {
	activeOnInit: PropTypes.string,
	className: PropTypes.string,
	items: PropTypes.arrayOf(PropTypes.shape({
		title: PropTypes.string,
		path: PropTypes.string,
	})),
	onClick: PropTypes.func,
};

export default Tabs;
