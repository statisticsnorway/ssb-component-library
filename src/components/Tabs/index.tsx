import React, { useState } from 'react'
import classNames from '../../utils/utils'

interface TabItem {
  title: string
  path: string
}

interface TabsProps {
  activeOnInit?: string
  className?: string
  items: TabItem[]
  onClick?: (path: string) => void
  id?: string
}

const Tabs: React.FC<TabsProps> = ({ activeOnInit = '', className, items, onClick = () => {}, id }) => {
  const [activeItem, setActiveItem] = useState(activeOnInit)

  const handleClick = (path: string) => {
    onClick(path)
    setActiveItem(path)
  }

  return (
    <div className={classNames('ssb-tabs', className)} role='tablist'>
      {items.map((item, i) => (
        <button
          className={`navigation-item ${activeItem === item.path ? 'active' : ''}`}
          onClick={() => handleClick(item.path)}
          key={item.path}
          role='tab'
          aria-selected={activeItem === item.path}
          aria-controls={`tabpanel-${i}-${id || ''}`}
        >
          <span>{item.title}</span>
        </button>
      ))}
    </div>
  )
}

export default Tabs
