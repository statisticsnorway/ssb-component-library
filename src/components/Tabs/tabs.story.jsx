import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';
import Divider from '../Divider';
import Tabs from './index';

const items = [
	{
		title: 'Get Started',
		path: 'get-started/',
	},
	{
		title: 'Components',
		path: 'components/',
	},
	{
		title: 'Templates',
		path: 'templates/',
	},
];

storiesOf('Tabs', module).addDecorator(centered)
	.add('Tabs', () => (
		<div>
			<Tabs items={items} />
			<Divider />
		</div>
	));
