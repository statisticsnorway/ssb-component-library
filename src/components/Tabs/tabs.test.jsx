import React from 'react';
import {shallow} from 'enzyme';
import Tabs from './index';

const items = [
	{
		title: 'Get Started',
		path: 'get-started/',
	},
	{
		title: 'Components',
		path: 'components/',
	},
	{
		title: 'Templates',
		path: 'templates/',
	},
];

describe('Tabs component', () => {
	test('Matches the snapshot', () => {
		const wrapper = shallow(<Tabs items={items}/>);
		expect(wrapper).toMatchSnapshot();
	});
	test('Sends callback', () => {
		const onClick = jest.fn();
		const wrapper = shallow(<Tabs onClick={onClick} items={items}/>);
		wrapper.find('.navigation-item').first().simulate('click');
		expect(onClick).toBeCalled();
	});
	test('Sets correct init state', () => {
		const wrapper = shallow(<Tabs items={items} activeOnInit={items[2].path}/>);
		expect(wrapper.find('.navigation-item').last().hasClass('active')).toEqual(true);
	});
});
