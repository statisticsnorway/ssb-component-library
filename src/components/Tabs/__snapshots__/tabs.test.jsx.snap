// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs component Matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ssb-tabs"
    role="tablist"
  >
    <button
      aria-controls="tabpanel-0-"
      aria-selected="false"
      class="navigation-item "
      role="tab"
    >
      <span>
        Get Started
      </span>
    </button>
    <button
      aria-controls="tabpanel-1-"
      aria-selected="false"
      class="navigation-item "
      role="tab"
    >
      <span>
        Components
      </span>
    </button>
    <button
      aria-controls="tabpanel-2-"
      aria-selected="false"
      class="navigation-item "
      role="tab"
    >
      <span>
        Templates
      </span>
    </button>
  </div>
</DocumentFragment>
`;
